codegen/gateway.go

```diff
@@ -578,12 +583,171 @@ func testFixtures(endpointConfigObj map[string]interface{}) (map[string]*Endpoin
+			sortedArr = append(sortedArr, k)
+		} else if v.Field.Required {
+			sortedArr = append(sortedArr, k)
+		}
+	}
+	sort.Strings(sortedArr)
+	return sortedArr
+}
+
+func resolveHeaders(
+	espec *EndpointSpec,
+	endpointConfigObj map[string]interface{},
+	key string,
+) error {
+	var (
+		keyMap = map[string]string{
+			reqHeaders: "http.req.metadata",
+			resHeaders: "http.res.metadata",
+		}
+		headersMap    = make(map[string]*TypedHeader)
+		headerModels  []string
+		annotationKey = keyMap[key]
+	)
+	defer func() {
+		if key == reqHeaders {
+			espec.ReqHeaders = headersMap
+		} else {
+			espec.ResHeaders = headersMap
+		}
+	}()
+	transformMap, err := loadHeadersFromConfig(endpointConfigObj, key)
+	if err != nil {
+		return err
+	}
+	method, err := findMethodByName(espec.ThriftMethodName, espec.ModuleSpec.Services)
+	if err != nil {
+		return err
+	}
+	for ak, av := range method.CompiledThriftSpec.Annotations {
+		if strings.HasSuffix(ak, annotationKey) {
+			headerModels = strings.Split(av, ",")
+			break
+		}
+	}
+	if len(headerModels) < 1 && len(transformMap) > 0 {
+		return errors.Errorf("header models %q unconfigured for transform", key)
+	}
+	if len(headerModels) < 1 {
+		return nil
+	}
+	for _, m := range headerModels {
+		typedHeaders, err := resolveHeaderModels(espec.ModuleSpec, m)
+		if err != nil {
+			return err
+		}
+		for hk, hv := range typedHeaders {
+			headersMap[hk] = hv
+		}
+	}
+	// apply header transform
+	for k, v := range transformMap {
+		typedHeader, ok := headersMap[k]
+		if !ok {
+			return errors.Errorf("unable to find header %q to transform", k)
+		}
+		typedHeader.TransformTo = v
+	}
+	return nil
+}
+
+func resolveHeaderModels(ms *ModuleSpec, modelPath string) (map[string]*TypedHeader, error) {
+	const (
+		headerPreix   = "headers"
+		httpRefSuffix = "http.ref"
+	)
+	loadModuleFromInclude := func(moduleName string) *compile.Module {
+		for pkgKey, pkg := range ms.CompiledModule.Includes {
+			if pkgKey == moduleName {
+				return pkg.Module
+			}
+		}
+		return nil
+	}
+	loadHeaderKeyFromField := func(field *compile.FieldSpec) *string {
+		for ak, av := range field.Annotations {
+			if avs := strings.Split(av, "."); len(avs) == 2 &&
+				avs[0] == headerPreix &&
+				strings.HasSuffix(ak, httpRefSuffix) {
+				headerKey := avs[1]
+				return &headerKey
+			}
+		}
+		return nil
+	}
+	loadHeadersFromCompiledModule := func(module *compile.Module, structName string) (map[string]*TypedHeader, error) {
+		var typeStruct *compile.StructSpec
+		var typedHeaders = make(map[string]*TypedHeader)
+		for tk, tv := range module.Types {
+			if ts, ok := tv.(*compile.StructSpec); tk == structName && ok {
+				typeStruct = ts
+				break
+			}
+		}
+		if typeStruct == nil {
+			return nil, errors.Errorf("unable to find typedHeaders %q", structName)
+		}
+		for _, field := range typeStruct.Fields {
+			headerKey := loadHeaderKeyFromField(field)
+			if headerKey == nil {
+				return nil, errors.New("unable to find header key")
```

Comment:
Can we have more informative message in this error? How do I know what's wrong when I get `unable to find header key`?